src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceClientService.java

```diff
@@ -0,0 +1,241 @@
+     * Base endpoint is in other words service (code) level endpoint.
+     * Each service has one base endpoint.
+     * Base endpoint has method '*' and path '**'.
+     *
+     * @param clientId
+     * @return
+     * @throws ClientNotFoundException
+     *
+     */
+    public List<ServiceClientDto> getServiceClientsByClient(ClientId clientId)
+            throws ClientNotFoundException {
+        ClientType clientType = clientRepository.getClient(clientId);
+        if (clientType == null) {
+            throw new ClientNotFoundException("Client " + clientId.toShortString() + " not found");
+        }
+
+        // Filter just acls that are set to base endpoints so they are on service code level
+        List<AccessRightType> serviceCodeLevelAcls = clientType.getAcl().stream()
+                .filter(acl -> acl.getEndpoint().isBaseEndpoint())
+                .collect(Collectors.toList());
+        List<AccessRightType> distinctAccessRightTypes = distinctAccessRightTypeByXroadId(serviceCodeLevelAcls);
+        return accessRightService.mapAccessRightsToServiceClients(clientType, distinctAccessRightTypes);
+    }
+
+    // Get unique AccessRightTypes from the given list
+    private List<AccessRightType> distinctAccessRightTypeByXroadId(List<AccessRightType> acls) {
+        HashMap<XRoadId, AccessRightType> uniqueServiceClientMap = new HashMap<>();
+        for (AccessRightType acl : acls) {
+            if (!uniqueServiceClientMap.containsKey(acl.getSubjectId())) {
+                uniqueServiceClientMap.put(acl.getSubjectId(), acl);
+            }
+        }
+        return new ArrayList(uniqueServiceClientMap.values());
+    }
+
+    /**
+     * Get single service client
+     *
+     * @param clientId
+     * @param dto
+     * @return
+     * @throws ClientNotFoundException if client with given id is not found
+     * @throws ServiceClientNotFoundException if service client with given parameters is not found
+     */
+    public ServiceClientDto getServiceClient(ClientId clientId, ServiceClientIdentifierDto dto)
+            throws ClientNotFoundException, ServiceClientNotFoundException {
+        List<ServiceClientDto> serviceClientsByClient = getServiceClientsByClient(clientId);
+        return serviceClientsByClient.stream()
+            .filter(scDto -> dto.isLocalGroup()
+                    ? dto.getLocalGroupId().toString().equals(scDto.getLocalGroupId())
+                    : dto.getXRoadId().equals(scDto.getSubjectId()))
+            .findFirst()
+            .orElseThrow(() -> {
+                String serviceClientIdentifier = dto.isLocalGroup()
+                        ? "Localgroup id: " + dto.getLocalGroupId()
+                        : "xRoadId: " + dto.getXRoadId().toShortString();
+                return new ServiceClientNotFoundException("Service client not found for ClientId: "
+                        + clientId.toShortString() + " and " + serviceClientIdentifier);
+            });
```

Comment:
Sonar claims this is not tested, maybe consider testing?

https://sonarqube.niis.org/component_measures?id=xroad&metric=new_coverage&pullRequest=464&selected=xroad%3Aproxy-ui-api%2Fsrc%2Fmain%2Fjava%2Forg%2Fniis%2Fxroad%2Frestapi%2Fservice%2FServiceClientService.java

Reply:
Added a test for testing searching a service client from a client that doesn't contain it.